cmd/docsync/main.go

```diff
@@ -0,0 +1,169 @@
+// Command docsync uploads a local directory to a specified
+// S3 bucket
+//
+// Usage:
+//
+//     docsync bucket bucketPrefix localDir
+//
+// where bucket is the name of your S3 bucket, bucketPrefix is the
+// S3 prefix to be applied to all uploaded files, and localDir is a
+// directory containing compiled docs (from docgenerate) or other
+// files to upload.
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"mime"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func main() {
+	bucket := os.Args[1]
+	bucketPrefix := os.Args[2]
+	localDir := os.Args[3]
+
+	localKeys := mustListContents(localDir)
+
+	region := "us-east-1"
+	sess := session.Must(session.NewSession(aws.NewConfig().WithRegion(region)))
+	svc := s3.New(sess)
+
+	var remoteKeys []string
+	err := svc.ListObjectsPages(&s3.ListObjectsInput{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(bucketPrefix),
+	}, func(page *s3.ListObjectsOutput, lastPage bool) bool {
+		for _, obj := range page.Contents {
+			if !(*obj.Key == "docs/") {
+				remoteKeys = append(remoteKeys, *obj.Key)
+			}
+		}
+		return true
+	})
+
+	if err != nil {
+		log.Fatalln("s3 list objects error:", err)
+	}
+
+	for _, k := range localKeys {
+		var body []byte
+
+		path := fmt.Sprintf("%s/%s", localDir, k)
+		body, err = ioutil.ReadFile(path)
+		if err != nil {
+			log.Fatalln(err.Error())
+		}
+
+		upload := &s3.PutObjectInput{
+			Bucket: aws.String(bucket),
+			Key:    aws.String(bucketPrefix + k),
+			Body:   bytes.NewReader(body),
+		}
+
+		ext := filepath.Ext(path)
+		contentType := mime.TypeByExtension(ext)
+
+		if contentType == "" {
+			upload.SetContentType("text/html")
+		} else {
+			upload.SetContentType(contentType)
+		}
+
+		_, err = svc.PutObject(upload)
+
+		if err != nil {
+			log.Fatalln(err.Error())
+		}
+	}
+
+	var prefixedLocalKeys []string
+	for _, k := range localKeys {
+		prefixedLocalKeys = append(prefixedLocalKeys, path.Join("docs", k))
```

Comment:
I think rather than using `path.Join` here, we should be prefixing with `bucketPrefix`. The setDiff is trying to see what is in the remote directory but not in the local directory, but in order for those things to be comparable, it needs to make the local keys (relative paths) look like remote keys ("absolute" paths with the prefix).